rhel/repositoryscanner.go

```diff
@@ -204,24 +206,66 @@ func (r *RepositoryScanner) Scan(ctx context.Context, l *claircore.Layer) (repos
 	if !ok || cm == nil {
 		return []*claircore.Repository{}, fmt.Errorf("rhel: unable to create a mappingFile object")
 	}
-	CPEs, err := mapContentSets(ctx, sys, cm)
+
+	// First, look at DNF.
+	repoids, err := dnf.FindRepoids(ctx, sys)
+	switch {
+	case err != nil: // continue to error check and return
+	case len(repoids) == 0:
+		// If DNF yielded nothing, check for content-sets data.
+		repoids, err = repoidsFromContentSets(ctx, sys)
+	default:
+	}
 	if err != nil {
 		return []*claircore.Repository{}, err
 	}
-	if CPEs == nil && r.apiFetcher != nil {
-		// Embedded content-sets are available only for new images.
-		// For old images, use fallback option and query Red Hat Container API.
-		ctx, done := context.WithTimeout(ctx, r.cfg.Timeout)
-		defer done()
-		CPEs, err = mapContainerAPI(ctx, sys, r.apiFetcher)
+
+	pairs := func(yield func(string, string) bool) {
+		for _, repoid := range repoids {
+			cpes, ok := cm.GetOne(ctx, repoid)
+			if !ok {
+				continue
+			}
+			for _, cpe := range cpes {
+				if !yield(repoid, cpe) {
+					return
+				}
+			}
+		}
+
+		if r.apiFetcher != nil {
+			// Embedded content-sets are unavailable in very old images.
+			// For these, use fallback option and query Red Hat Container API.
+			ctx, done := context.WithTimeout(ctx, r.cfg.Timeout)
+			defer done()
+
+			// !!! This ends up being a weird nonlocal return. !!!
```

Comment:
what does this mean?

Reply:
I think it's saying that any error returned from the `mapContainerAPI` is not evaluated or returned until outside of this function body L#259